python/pyarrow/_s3fs.pyx

```diff
@@ -74,6 +74,13 @@ cdef class S3FileSystem(FileSystem):
         Whether to connect anonymously if access_key and secret_key are None.
         If true, will not attempt to look up credentials using standard AWS
         configuration methods.
+    use_web_identity: boolean, default False
+        Whether to connect using an assumed role authenticated using
+        a web identity token. The required settings are derived from
+        environment variables such as AWS_ROLE_ARN,
+        AWS_WEB_IDENTITY_TOKEN_FILE and AWS_ROLE_SESSION_NAME.
+        If true, will not attempt to look up credentials using other
+        AWS configuration methods.
```

Comment:
Hmm... I think it would be better to avoid adding boolean arguments for each different kind of authentication.
I'm not sure what a good API would be, but perhaps we could add a generic `auth` parameter that would accept the following values:
* `auth='default'`
* `auth='anonymous'`
* `auth={'access_key': '...', 'secret_key': '...', 'session_token': '...'}`
* `auth={'role_arn': '...', 'session_name': '...', 'external_id': '...', 'load_frequency': 123}`
* `auth='web_identity'`

@jorisvandenbossche What do you think?

Reply:
Without knowing too much about S3 / AWS auth, your proposal feels quite natural for expressing the different options, and could be a nice alternative to a bunch of partially exclusive keyword arguments each for a different way of authorization. We would (initially) need to keep the keywords for backward compatibility.

Reply:
Whatever you decide, could you please document and write up a JIRA for doing the same in the R package (assuming you're not going to go all the way and update the R bindings in this PR, though you're more than welcome to :)

Reply:
@pitrou That makes sense to me as well.
I think it might be better to do that in a separate PR though (to keep the scope of this one limited).
Let me know.

Reply:
I would be ok with doing the Python changes in a separate PR, but then you should remove them from this PR ;-)

Reply:
@sahil1105 you could also do a pre-cursor PR to refactor the existing options like discussed here

Reply:
Wouldn't we still need the python changes in this PR even if we move to using a generic `auth` parameter (since we need to keep the keywords around for now for backward compatibility)? Or am I missing something?

Reply:
The Python changes in this PR are not needed for backwards compatibility? (since those are new options) 
So assuming we refactor the python options to use a generic `auth` parameter, then this PR with new functionality should directly use that new parameter, and not introduce a new `use_web_identity` keyword that would then later need to be deprecated. 
So for this PR, I think either remove any Python bindings in this PR (and keep this PR to only add the C++ functionality, and later do a follow-up PR to add the python bindings) or either first do a precursor PR to refactor the options to use `auth` (and then afterwards this PR can be updated to add the web identity functionality into an `auth` keyword).

Reply:
I agree about removing the `use_web_identity` keyword arg. I meant all the other Python changes. Those are actually just simple error-checking changes that we should keep in my opinion.
Related Q: Should I remove `S3CredentialsKind` and related changes? I think it's good to have them, but I'll defer to you regarding whether to keep them.

Reply:
I agree that it's ok to keep the additional error-checking changes, as well as the `S3CredentialsKind` declaration.

Reply:
Done. Sorry for the delay.